tensorwaves/interfaces.py

```diff
@@ -0,0 +1,102 @@
+"""Defines top level interfaces of tensorwaves.
+
+"""
+
+from abc import ABC, abstractmethod
+
+
+class Function(ABC):
+    """
+    Defines the interface of a callable function.
+    The parameters of the model are separated from the domain variables. This
+    follows the mathematical definition, in which a function defines its domain
+    and parameters. However specific points in the domain are not relevant.
+    Hence while the domain variables are the argument of the evaluation
+    (see :func:`~Function.__call__`), the parameters are controlled via a
+    getter and setter (see :func:`~Function.parameters`). The reason for this
+    separation is to facilitate the events when parameters have changed.
+    """
+    @abstractmethod
+    def __call__(self, dataset: dict) -> list:
+        """Evaluate the function.
+
+        Args:
+            dataset: a dictionary with domain variable names as keys and a
+            an list or `numpy.array` of  with a
+
+        Return:
+            list or :class:`numpy.array` of values
+        """
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        """
+        """
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        """
+        """
+
+
+class Estimator(ABC):
+    @abstractmethod
+    def __call__(self) -> float:
+        pass
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        pass
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        pass
+
+
+class Kinematics(ABC):
+    @abstractmethod
+    def convert(self, events: dict) -> dict:
+        pass
+
+    @abstractmethod
+    def reduce_to_phase_space(self, events: dict) -> dict:
+        pass
+
+    @property
+    @abstractmethod
+    def phase_space_volume(self) -> float:
+        pass
+
+
+class Optimizer(ABC):
+    @abstractmethod
+    def optimize(self, estimator: Estimator, initial_parameters: dict) -> dict:
+        pass
+
+
+class PhaseSpaceGenerator(ABC):
```

Comment:
Is a phase space generator independent of a uniform generator?

Reply:
Yeah, the idea here is that we want to have full reproducability or deterministic behavior. So if you chose a specific phase space generator and seed, the output should always be the same. Since you need the random number generation in the generation of phase space events and in possible hit & miss / accept and reject, it was best to hand a random generator to the phase space generator. That avoids mixups (having two random generators and maintaining their state) and also gives flexibility.

Reply:
See https://github.com/ComPWA/tensorwaves/pull/30/files/a2d859cf1dfd332e6e18f4f97a2bad5faea54911#r413077761
Or should the `random_generator` in `generate` be a `seed` (`int` or `float`)?